File: куча.py
# Бинарная куча
class BinaryHeap:
    def __init__(self):
        self.heap = []
    
    def parent(self, i):
        return (i - 1) // 2

# Биномиальная куча
class BinomialTree:
    def __init__(self, key):
        self.key = key
        self.children = []

# Куча Фибоначчи 
class FibonacciHeap:
    def __init__(self):
        self.min_node = None
        self.trees = []

File: куча.cpp
// Бинарная куча 
template<typename T>
class BinaryHeap {
private:
    std::vector<T> data;
    void heapify_up(int index);
};

// Биномиальная куча 
template<typename T>
class BinomialHeap {
    struct Node {
        T value;
        std::vector<Node*> children;
    };
    std::vector<Node*> trees;
};

// Куча Фибоначчи 
template<typename T>
class FibonacciHeap {
    struct Node {
        T key;
        Node* parent;
        std::list<Node*> children;
    };
    Node* min_node;
};